### PR TITLE
feat: support the async approval flow

### PR DESCRIPTION
**Pull request by owencraston**
resolves: https://github.com/MetaMask/accounts-planning/issues/12

## Snap:
  - Added a new `pendingRequests` field to the `SimpleKeyring` class ([code](https://github.com/MetaMask/snap-simple-keyring/blob/d3f7f0156c59059c995fea87f90a3d0ad3a4c135/packages/snap/src/keyring.ts#L43))
  - Modified the return of the `submitRequest` method to:
    - Added the incoming request to `pendingRequests`
    - Return `{ pending: true }` ([code](https://github.com/MetaMask/snap-simple-keyring/blob/d3f7f0156c59059c995fea87f90a3d0ad3a4c135/packages/snap/src/keyring.ts#L167-L170))
  - Implemented the [`approveRequest`](https://github.com/MetaMask/snap-simple-keyring/blob/d3f7f0156c59059c995fea87f90a3d0ad3a4c135/packages/snap/src/keyring.ts#L173) and [`rejectRequest`](https://github.com/MetaMask/snap-simple-keyring/blob/d3f7f0156c59059c995fea87f90a3d0ad3a4c135/packages/snap/src/keyring.ts#L179) methods
    - The `approveRequest` method calls the `snap_manageAccounts` RPC method ([example](https://github.com/MetaMask/snap-simple-keyring/blob/d3f7f0156c59059c995fea87f90a3d0ad3a4c135/packages/snap/src/keyring.ts#L93-L99)) with the following `params`:

      ```ts
           await snap.request({
              method: 'snap_manageAccounts',
              params: {
                method: 'submitResponse',
                params: { id: _id, result: signature },
              },
            });
      ```

    - The `rejectRequest` method calls the `snap_manageAccounts` RPC method ([example](https://github.com/MetaMask/snap-simple-keyring/blob/d3f7f0156c59059c995fea87f90a3d0ad3a4c135/packages/snap/src/keyring.ts#L93-L99)) with the following `params`:

      ```ts
            await snap.request({
              method: 'snap_manageAccounts',
              params: {
                method: 'submitResponse',
                params: { id: _id, result: null },
              },
            });
      ```
    - Removed the `requests` field in favour of pending requests
    
## Site

- The `listRequests` was already wired up
- the `approveRequest` request was already wired up
- Modified the reject component to call my new `rejectRequest` function

## Testing
Screenshot 2023-08-29 at 10 05 27 AM

https://github.com/MetaMask/snap-simple-keyring/assets/22918444/10190283-37fd-48c1-a9a3-958f5a368fa9

Assuming you have the latest Metamask flask extension built and installed in the browser here are the testing steps.

#### Testing Async Requests
1. make sure you have imported an account with testnet eth
2. run yarn start in root directory
3. modify the defaultSnapOrigin in `packages/site/src/config/snap.ts` to be `export const defaultSnapOrigin = 'local:http://localhost:8080';`
4. open http://localhost:8000/
5. make sure that `Use Synchronous Approval` is toggled off
6. click `List Accounts` in the snap keyring dapp. `Notice` there are no accounts
7. click create account and execute method. An error will be thrown but this is alright. check metamask extension and you should see a new snap account appear in the accounts list.
8. Click `List Accounts` again and you should see the account response
9. fund your new snap account with testnet eth
10. once the snap account has funds, send a transaction from that account. It should remain pending.
11. go to the dapp, click `List Requests` and you should see a request object
12. extract the `id` from the response and paste it into the `Approve request` field and click execute.
13. Navigate to the extension and that pending transaction should now be confirmed
14. list the requests again and the request should no longer be in the pending request list
15. repeat steps 9 and 10
16. extract the new request id and paste it into the `Reject request` field
17. Navigate to the extension and that latest transaction should have failed.
18. 14. list the requests again and the request should no longer be in the pending request list

#### Testing Synchronous approvals
- assuming you have a snap account that has been funded from the previous steps you can test the sync flow as follows

1. Turn the `Use Synchronous Approval` toggle on, you should see a green checkmark
2. go into metamask and within your snap account perform a send transaction
3. once you confirm the transaction, wait a few seconds and the transaction should be confirmed
4. clicking list requests should result in an empty array
5. refresh the snap simple keyring dapp and the toggle should still be set to true
6. toggle the Synchronous approvals off and then go through the async approval steps mentioned above
7. confirm that you are using the async flow that requires manual approval with the form

### CODE REVIEW

File: packages/snap/snap.manifest.json [hunk 1]
```diff
@@ -7,7 +7,7 @@
     "url": "git+https://github.com/MetaMask/snap-simple-keyring.git"
   },
   "source": {
-    "shasum": "s5XnCGE4PLuRg5wcW7QyK3a+FE8P3zfJLemiz6t86O4=",
+    "shasum": "KvlHpqgEHTR53aD3rwHq9e5+9ep25+pjEhEBFmAgFes=",
```
Comment (owencraston) [inline packages/snap/snap.manifest.json:10]: should these changes be committed?

Comment (gantunesr) [reply packages/snap/snap.manifest.json:10]: Yes, if the source code changes the shasum should also change

**Comment by gantunesr** [comment]
From stand up:

@danroc and @owencraston synced and it was decided to add a toggle to switch between sync and async flows

**Comment by owencraston** [comment]
Here is a video of me using async and sync approvals/rejections in the same dapp
https://github.com/MetaMask/snap-simple-keyring/assets/22918444/3e22ea98-5f11-4fb3-ac1c-0f710d772484

**Review by HowardBraham — APPROVED** [review]

**Review by HowardBraham — APPROVED** [review]

File: packages/snap/src/keyring.ts [hunk 2]
```diff
@@ -176,24 +188,81 @@ export class SimpleKeyring implements Keyring {
...
-      pending: false,
-      result: signature,
-    };
+    if (this.#useSynchronousApprovals) {
+      return this.#handleSynchronousSubmitRequest(request);
+    }
+    return this.#handleAsyncSubmitRequest(request);
   }
 
   async approveRequest(_id: string): Promise<void> {
```
Comment (danroc) [inline packages/snap/src/keyring.ts:197]: This variable has the `_` prefix because it wasn't used, should be fine to rename it to `id`.

**Review by danroc — APPROVED** [review]

**Comment by owencraston** [comment]
@danroc I have went ahead and added a success message to the approve and reject buttons. I also show the error if there is one.
Screenshot 2023-08-29 at 1 43 24 PM

Screenshot 2023-08-29 at 1 45 28 PM

Also when you toggle the async vs sync mode, the account list does not disappear.

**Review by HowardBraham — APPROVED** [review]

**Review by HowardBraham — APPROVED** [review]

**Pull request merged**